src/server/hset_family.cc

```diff
@@ -1020,24 +1048,61 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
     StringVec str_vec;
 
     if (pv.Encoding() == kEncodingStrMap2) {
-      // TODO: to create real random logic.
       StringMap* string_map = (StringMap*)pv.RObjPtr();
-
-      sds key = string_map->begin()->first;
-      str_vec.emplace_back(key, sdslen(key));
+      if (args.size() == 1) {
+        auto [key, value] = string_map->RandomPair();
+        str_vec.emplace_back(key, sdslen(key));
+      } else {
+        size_t actual_count =
+            (count >= 0) ? std::min(size_t(count), string_map->Size()) : abs(count);
+        std::vector<sds> keys, vals;
+        if (count >= 0) {
+          string_map->RandomPairsUnique(actual_count, keys, vals, with_values);
+        } else {
+          string_map->RandomPairs(actual_count, keys, vals, with_values);
+        }
+        for (size_t i = 0; i < actual_count; ++i) {
+          str_vec.emplace_back(keys[i], sdslen(keys[i]));
+          if (with_values) {
+            str_vec.emplace_back(vals[i], sdslen(vals[i]));
+          }
+        }
+      }
     } else if (pv.Encoding() == kEncodingListPack) {
       uint8_t* lp = (uint8_t*)pv.RObjPtr();
       size_t lplen = lpLength(lp);
       CHECK(lplen > 0 && lplen % 2 == 0);
-
       size_t hlen = lplen / 2;
-      listpackEntry key;
-
-      lpRandomPair(lp, hlen, &key, NULL);
-      if (key.sval) {
-        str_vec.emplace_back(reinterpret_cast<char*>(key.sval), key.slen);
+      if (args.size() == 1) {
+        listpackEntry key;
+        lpRandomPair(lp, hlen, &key, NULL);
+        StrVecEmplaceBack(str_vec, key);
       } else {
-        str_vec.emplace_back(absl::StrCat(key.lval));
+        size_t actual_count = (count >= 0) ? std::min(size_t(count), hlen) : abs(count);
+        listpackEntry *keys = nullptr, *vals = nullptr;
+        keys = new listpackEntry[actual_count];
```

Comment:
Oh wait, I did not see this. Plz use `absl::Cleanup`

Reply:
You delete this later on with `[]`. This is no RAII and raw memory managements and it's looking for trouble

Reply:
I will use unique_ptr